pkg/log/locallogger/locallogger.go

```diff
@@ -69,5 +194,122 @@ func filterResults(keyvals ...interface{}) {
 			}
 		}
 	}
+}
+
+// hashKeyvals creates a hash of the key-value pairs for deduplication
+// We exclude timestamp and caller fields since they change for identical log content
+func (ll *localLogger) hashKeyvals(keyvals ...interface{}) string {
+	return hashKeyValuePairs(keyvals...)
+}
+
+// hashKeyValuePairs creates a consistent hash from key-value pairs, excluding timestamp fields
+func hashKeyValuePairs(keyvals ...interface{}) string {
+	hasher := sha256.New()
+
+	// Collect non-excluded key-value pairs
+	var pairs []string
+	for i := 0; i < len(keyvals); i += 2 {
+		if i+1 >= len(keyvals) {
+			break
+		}
+
+		key := fmt.Sprintf("%v", keyvals[i])
+
+		// Skip excluded fields for content-based deduplication
+		if excludedHashFields[key] {
+			continue
+		}
+
+		value := fmt.Sprintf("%v", keyvals[i+1])
+		pairs = append(pairs, key+":"+value)
+	}
+
+	// Sort pairs for consistent hashing
+	sort.Strings(pairs)
+
+	// Hash the sorted pairs
+	for _, pair := range pairs {
+		hasher.Write([]byte(pair + ";"))
+	}
+
+	return fmt.Sprintf("%x", hasher.Sum(nil))
+}
+
+// shouldSkipDuplicate checks if this log message is a duplicate and should be skipped
+// Returns (skip, duplicateCount) where skip indicates if the message should be skipped
+// and duplicateCount is the number of times this message has been seen
+func (ll *localLogger) shouldSkipDuplicate(hash string) (bool, int) {
+	ll.dedupMutex.Lock()
+	defer ll.dedupMutex.Unlock()
+
+	now := time.Now()
+
+	// Perform cleanup if needed
+	if now.Sub(ll.lastCleanup) > cleanupInterval {
+		ll.cleanupCacheUnsafe(now)
+		ll.lastCleanup = now
+	}
 
+	entry, exists := ll.dedupCache[hash]
+	if !exists {
+		// First time seeing this message
+		ll.dedupCache[hash] = &logEntry{
+			firstSeen:  now,
+			lastSeen:   now,
+			count:      1,
+			lastLogged: now,
+		}
+		return false, 1 // Don't skip, log it (first occurrence)
+	}
+
+	// Update the existing entry
+	entry.lastSeen = now
+	entry.count++
+
+	// Check if enough time has passed since we last logged this message
+	// We use a simple strategy: log the duplicate if it's been more than the interval
+	// since we last logged this exact message, and include the count
+	if now.Sub(entry.lastLogged) > duplicateLogInterval {
+		entry.lastLogged = now
+		return false, entry.count // Don't skip, log it with count
+	}
+
+	// Skip this duplicate
+	return true, entry.count
+}
+
+// cleanupCacheUnsafe removes expired entries from the deduplication cache
+// Must be called with dedupMutex held
```

Comment:
I think we should instead lock `dedupMutex` inside this function (we'd also have to update `shouldSkipDuplicate` to wait to lock `dedupMutex` until after it's finished calling `cleanupCacheUnsafe`). That way, we can guarantee that this function is handling the cache safely, instead of relying on the caller not making a mistake with the mutex.